### PR TITLE
[cpe/alps] mpi.py

### PR DESCRIPTION
**Pull request by jgphpc**
epic https://jira.cscs.ch/browse/VCMSA-263
+ https://jira.cscs.ch/browse/VCMSA-339

### CODE REVIEW

**Comment by jgphpc** [comment]
[ci](https://gitlab.com/cscs-ci/ci-testing/webhook-ci/mirrors/6557018579987861/835515875116200/-/jobs/4343509153#L277): 🆗

File: ci/cpe.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  tags: ['hohgant-login-baremetal']
+  variables:
+    GIT_STRATEGY: fetch
+  script:
+    - rm -fr venv_reframe_cpe
+    - python3 -m venv venv_reframe_cpe
+    - source venv_reframe_cpe/bin/activate
+    - pip install reframe-hpc
+    - export RFM_AUTODETECT_XTHOSTNAME=1
+    - export RFM_REMOTE_DETECT=0
```
Comment (ekouts) [inline ci/cpe.yml:19]: Why are do we skip the autodetection of the remote partitions? Isn't it supposed to be used in our generic tests?

File: checks/prgenv/mpi.py [hunk 2]
```diff
@@ -53,16 +31,23 @@ class MpiInitTest(rfm.RegressionTest):
...
     @run_after('init')
     def set_cpp_flags(self):
         self.build_system.cppflags = self.cppflags[self.required_thread]
 
+    @run_after('setup')
+    def skip_if_no_mpi(self):
+        self.skip_if(self.current_partition.name.startswith('login'),
+                     'skip login partition')
+        self.skip_if(self.current_environ.name.startswith('builtin'),
+                     'skip builtin pe')
```
Comment (ekouts) [inline checks/prgenv/mpi.py:46]: Same for the environment. I would prefer to add a feature for all the mpi envs.

Comment (jgphpc) [reply checks/prgenv/mpi.py]: https://github.com/eth-cscs/cscs-reframe-tests/pull/77/commits/6131d0eb876873aa4bfe70de4cd91597642f3f79

File: checks/prgenv/mpi.py [hunk 3]
```diff
@@ -53,16 +31,23 @@ class MpiInitTest(rfm.RegressionTest):
...
     tags = {'production', 'craype'}
 
     @run_after('init')
     def set_cpp_flags(self):
         self.build_system.cppflags = self.cppflags[self.required_thread]
 
+    @run_after('setup')
+    def skip_if_no_mpi(self):
+        self.skip_if(self.current_partition.name.startswith('login'),
+                     'skip login partition')
```
[the remark this answers is not in the document]
Comment (jgphpc) [reply checks/prgenv/mpi.py]: https://github.com/eth-cscs/cscs-reframe-tests/pull/77/commits/6131d0eb876873aa4bfe70de4cd91597642f3f79

File: checks/prgenv/mpi.py [hunk 4]
```diff
@@ -11,35 +11,13 @@
...
     '''
     required_thread = parameter(['single', 'funneled', 'serialized',
                                  'multiple'])
-    valid_prog_environs = ['PrgEnv-aocc', 'PrgEnv-cray', 'PrgEnv-gnu',
-                           'PrgEnv-intel', 'PrgEnv-pgi', 'PrgEnv-nvidia',
-                           'PrgEnv-nvhpc']
-    valid_systems = ['daint:gpu', 'daint:mc', 'dom:gpu', 'dom:mc', 'eiger:mc',
-                     'pilatus:mc', 'hohgant:nvgpu', 'hohgant:gpu-squashfs']
+    valid_prog_environs = ['*']
+    valid_systems = ['*']
```
Comment (ekouts) [inline checks/prgenv/mpi.py:20]:
```suggestion
    valid_prog_environs = ['+mpi']
    valid_systems = ['+mpi']
```

File: checks/prgenv/mpi.py [hunk 5]
```diff
@@ -53,16 +31,23 @@ class MpiInitTest(rfm.RegressionTest):
...
     def set_cpp_flags(self):
         self.build_system.cppflags = self.cppflags[self.required_thread]
 
+    @run_after('setup')
+    def skip_if_no_mpi(self):
+        self.skip_if('mpi' not in self.current_partition.features,
+                     'skip partition with no mpi')
+        self.skip_if('mpi' not in self.current_environ.features,
+                     'skip environ with no mpi')
+
```
Comment (ekouts) [inline checks/prgenv/mpi.py:47]: We can remove this and just add it in the `valid_prog_environs`, `valid_systems`.

File: checks/prgenv/cuda/nvml_check.py [hunk 6]
```diff
@@ -9,42 +9,38 @@
...
-        'cp $CUDATOOLKIT_HOME/nvml/example/example.c .',
-        'patch -i ./nvml_example.patch'
-    ]
-    maintainers = ['AJ', 'SK']
+    sourcepath = '$CUDATOOLKIT_HOME/nvml/example/example.c'
+    maintainers = []
     tags = {'production', 'craype', 'external-resources', 'health'}
 
+    @run_before('compile')
+    def setup_nvhpc(self):
```
Comment (jgphpc) [inline checks/prgenv/cuda/nvml_check.py:22]: @ekouts : are you ok with that ?

File: config/systems/hohgant.py [hunk 7]
```diff
@@ -162,27 +162,27 @@
...
-            'modules': ['cray', 'PrgEnv-intel']
+            'modules': ['cray', 'PrgEnv-gnu'],
+            'features': ['mpi', 'cuda']
         },
         {
             'name': 'PrgEnv-nvhpc',
             'target_systems': ['hohgant'],
             'modules': ['cray', 'PrgEnv-nvhpc'],
+            'features': ['mpi', 'cuda'],
+            # 'prepare_cmds': ['module unload cudatoolkit'],
```
Comment (ekouts) [inline config/systems/hohgant.py:185]: This line is not needed right?

**Review by ekouts — CHANGES_REQUESTED** [review]
I have a very small comment, otherwise lgtm and we can merge after

**Review by ekouts — APPROVED** [review]

**Pull request merged**